src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,164 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final double ARRAY_INCREMENT_FACTOR = 1.5;
+    private static final int DEFAULT_ARRAY_SIZE = 10;
+
+    private Object[] elementArray;
+    private int lastArrayIndex;
+
+    public ArrayList() {
+        this.elementArray = new Object[DEFAULT_ARRAY_SIZE];
+    }
+
     @Override
     public void add(T value) {
-
+        if (lastArrayIndex == elementArray.length) {
+            elementArray = grow(elementArray, elementArray.length);
+        }
+        elementArray[lastArrayIndex] = value;
+        lastArrayIndex++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index >= 0 && index <= lastArrayIndex) {
+            if (lastArrayIndex < elementArray.length) {
+                elementArray = addValueByIndex(elementArray, index, value);
+            } else {
+                elementArray = grow(elementArray, elementArray.length);
+                elementArray = addValueByIndex(elementArray, index, value);
+            }
+            lastArrayIndex++;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index "
+                    + index
+                    + " out of bounds for length "
+                    + (lastArrayIndex - 1)
+            );
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int newArraySize;
+        if (list != null && !list.isEmpty()) {
+            if (elementArray.length - lastArrayIndex < list.size()) {
+                newArraySize = elementArray.length + list.size() - lastArrayIndex;
+                elementArray = grow(elementArray, newArraySize);
+            }
+            for (int i = 0; i < list.size(); i++) {
+                elementArray[lastArrayIndex] = list.get(i);
+                lastArrayIndex++;
+            }
+        }
```

Comment:
I cannot simplify this method, or partially transfer the logic, because this method has its own logic. we also talked about it on QnA , and it was accepted